netket/graph/_lattice_edge_logic.py

```diff
@@ -0,0 +1,188 @@
+
+    basis_coords accepted as an array including sublattice coordinate in last column
+    or as a tuple of unit cell coordinates and a shared sublattice index
+    """
+    if isinstance(basis_coords, Tuple):
+        basis_coords, sl = basis_coords
+    else:
+        basis_coords, sl = basis_coords[:, :-1], basis_coords[:, -1]
+
+    # Accepts extended shells
+    basis_coords = basis_coords % extent
+
+    # Index difference between sites one lattice site apart in each direction
+    # len(site_offsets) for the last axis, as all sites in one cell are listed
+    # factor of extent[-1] to the penultimate axis, etc.
+    radix = np.cumprod([len(site_offsets), *extent[:0:-1]])[::-1]
+
+    return basis_coords @ radix + sl
+
+
+# Near-neighbour search logic
+
+
+def create_padded_sites(basis_vectors, extent, site_offsets, pbc, order):
+    """Generates all lattice sites in an extended shell"""
+    extra_shells = np.where(pbc, order, 0)
+    basis_coords, positions = create_site_positions(
+        basis_vectors, extent, site_offsets, extra_shells
+    )
+    equivalent_ids = site_to_idx(basis_coords, extent, site_offsets)
+
+    return positions, equivalent_ids
+
+
+def get_naive_edges(positions, cutoff, order):
+    """
+    Given an array of spatial `positions`, returns a list `es`, so that
+    `es[k]` contains all pairs of (k + 1)-nearest neighbors up to `order`.
+    Only edges up to distance `cutoff` are considered.
+    """
+    kdtree = cKDTree(positions)
+    dist_matrix = kdtree.sparse_distance_matrix(kdtree, cutoff)
+    row, col, dst = find(triu(dist_matrix))
+    dst = comparable(dst)
+    _, ii = np.unique(dst, return_inverse=True)
+
+    return [sorted(list(zip(row[ii == k], col[ii == k]))) for k in range(order)]
+
+
+def get_nn_edges(
+    basis_vectors,
+    extent,
+    site_offsets,
+    pbc,
+    distance_atol,
+    order,
+):
+    positions, ids = create_padded_sites(
+        basis_vectors, extent, site_offsets, pbc, order
+    )
+    naive_edges_by_order = get_naive_edges(
+        positions,
+        order * np.linalg.norm(basis_vectors, axis=1).max() + distance_atol,
+        order,
+    )
+    colored_edges = []
+    for k, naive_edges in enumerate(naive_edges_by_order):
+        true_edges = set()
+        for node1, node2 in naive_edges:
+            # switch to real node indices
+            node1 = ids[node1]
+            node2 = ids[node2]
+            if node1 == node2:
+                raise RuntimeError(
+                    f"Lattice contains self-referential edge {(node1, node2)} of order {k}"
+                )
+            elif node1 > node2:
+                node1, node2 = node2, node1
+            true_edges.add((node1, node2))
+        for edge in true_edges:
+            colored_edges.append((*edge, k))
+    return colored_edges
+
+
+# Unit cell distribution logic
+
+
+def get_custom_edges(basis_vectors, extent, site_offsets, pbc, atol, descriptor):
+    def translated_edges(sl1, sl2, distance, color):
```

Comment:
i prefer input validation to be done at the beginning...
also import `from textwrap import dedent`
```suggestion
    if not all([len(desc) in (3,4) for desc in descriptor])
        raise ValueError(dedent(
            """
            The custom edges descriptor list must contain tuples of length 3 or 4.
            Every tuple must contain two sublattice indices (integers), a distance vector
            and can optionally include an integer to represent the color of that edge.
            
            Check the docstring of `nk.graph.Lattice` for more informations.
            """
        ))
    
    def translated_edges(sl1, sl2, distance, color):
```

Reply:
```suggestion
    if not all([len(desc) in (3,4) for desc in descriptor])
        raise ValueError(dedent(
            """
            custom_edges must be a list of tuples of length 3 or 4.
            Every tuple must contain two sublattice indices (integers), a distance vector
            and can optionally include an integer to represent the color of that edge.
            
            Check the docstring of `nk.graph.Lattice` for more informations.
            """
        ))
    
    def translated_edges(sl1, sl2, distance, color):
```